File: src/game/quests.ts
import type { QuestTemplate } from './types.js';

export interface QuestProgress {
  questId: string;
  status: 'active' | 'complete' | 'turned_in';
  progress: Record<string, number>;
}

export function createQuestProgress(quest: QuestTemplate): QuestProgress {
  const progress: Record<string, number> = {};
  for (const obj of quest.objectives) {
    progress[obj.target] = 0;
  }
  return { questId: quest.id, status: 'active', progress };
}

export function checkQuestComplete(quest: QuestTemplate, qp: QuestProgress): boolean {
  return quest.objectives.every((obj) => (qp.progress[obj.target] ?? 0) >= obj.count);
}

export function formatQuestProgress(quest: QuestTemplate, qp: QuestProgress): string {
  const lines = [`[${quest.name}] ${quest.description}`, 'Objectives:'];
  for (const obj of quest.objectives) {
    const current = qp.progress[obj.target] ?? 0;
    const done = current >= obj.count;
    lines.push(`  ${done ? '✓' : '○'} ${obj.description} (${current}/${obj.count})`);
  }
  return lines.join('\n');
}

File: crates/realm-core/src/quests.rs
use crate::types::QuestTemplate;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum QuestStatus {
    Active,
    Complete,
    #[serde(rename = "turned_in")]
    TurnedIn,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QuestProgress {
    #[serde(rename = "questId")]
    pub quest_id: String,
    pub status: QuestStatus,
    pub progress: HashMap<String, i32>,
}

pub fn create_quest_progress(quest: &QuestTemplate) -> QuestProgress {
    let progress = quest
        .objectives
        .iter()
        .map(|obj| (obj.target.clone(), 0))
        .collect();
    QuestProgress {
        quest_id: quest.id.clone(),
        status: QuestStatus::Active,
        progress,
    }
}

pub fn check_quest_complete(quest: &QuestTemplate, qp: &QuestProgress) -> bool {
    quest.objectives.iter().all(|obj| {
        qp.progress.get(&obj.target).copied().unwrap_or(0) >= obj.count
    })
}

pub fn format_quest_progress(quest: &QuestTemplate, qp: &QuestProgress) -> String {
    let mut lines = vec![
        format!("[{}] {}", quest.name, quest.description),
        "Objectives:".to_string(),
    ];
    for obj in &quest.objectives {
        let current = qp.progress.get(&obj.target).copied().unwrap_or(0);
        let done = current >= obj.count;
        let marker = if done { "✓" } else { "○" };
        lines.push(format!(
            "  {marker} {} ({current}/{})",
            obj.description, obj.count
        ));
    }
    lines.join("\n")
}